File: Model/MarkerAction.m
//
//  MarkerAction.m
//  aestheticodes
//
//  Created by Kevin Glover on 03/07/2014.
//  Copyright (c) 2014 Horizon. All rights reserved.
//
#import "MarkerAction.h"

@implementation NSDictionary (Primitive)

-(NSString*)stringForKey:(NSString*)key withDefault:(NSString*)value
{
	NSString* string = self[key];
	if(string != nil)
	{
		return string;
	}
	return value;
}

-(BOOL)boolForKey:(NSString*)key withDefault:(bool)value
{
	NSNumber* number = self[key];
	if(number != nil)
	{
		return [number boolValue];
	}
	return value;
}

@end

@implementation MarkerAction

@synthesize description;

-(instancetype)init
{
	self = [super init];
	if(self)
	{
		self.editable = true;
		self.visible = true;
		self.showDetail = true;
	}
	return self;
}

-(NSDictionary*)toDictionary
{
	NSMutableDictionary* result = [[NSMutableDictionary alloc] init];
	
	[result setValue:self.code forKey:@"code"];
	[result setValue:self.title forKey:@"title"];
	[result setValue:self.description forKey:@"description"];
	[result setValue:self.action forKey:@"action"];
	[result setValue:self.image forKey:@"image"];
	
	[result setValue:[NSNumber numberWithBool:self.editable] forKey:@"editable"];
	[result setValue:[NSNumber numberWithBool:self.visible] forKey:@"visible"];
	[result setValue:[NSNumber numberWithBool:self.showDetail] forKey:@"showDetail"];
	
	return result;
}

+(BOOL)propertyIsOptional:(NSString*)propertyName
{
	return YES;
}

-(void)load:(NSDictionary *)data
{
	self.code = data[@"code"];
	self.title = [data valueForKey:@"title"];
	self.description = [data valueForKey:@"description"];
	self.action = [data valueForKey:@"action"];
	self.image = [data valueForKey:@"image"];
	
	self.editable = [data boolForKey:@"editable" withDefault:true];
	self.visible = [data boolForKey:@"visible" withDefault:true];
	self.showDetail	= [data boolForKey:@"showDetail" withDefault:true];
}

@end

File: Model/MarkerAction.swift
//
//  MarkerDetail.swift
//  aestheticodes
//
//  Created by Kevin Glover on 12/06/2014.
//  Copyright (c) 2014 Horizon. All rights reserved.
//

import Foundation

class MarkerAction
{
	var code: String
	var action: String?
	var editable = true
	var showDetail = true
	var description: String?
	var title = "Marker %@"
	var image: String?
	var visible = true
	
	init(code: String)
	{
		self.code = code
	}
	
	init(json: JSONValue)
	{
		code = json["code"].string!
		NSLog("Adding marker \(code)")
		action = json["action"].string
		if json["editable"].bool
		{
			editable = json["editable"].bool!
		}

		NSLog("\(action)")
		
		if json["title"].string
		{
			title = json["title"].string!
		}
		
		if json["showDetail"].bool
		{
			showDetail = json["showDetail"].bool!
		}
		
		if json["visible"].bool
		{
			visible = json["visible"].bool!
		}
		
		description = json["description"].string
		image = json["image"].string
	}
	
	init(dict: NSDictionary)
	{
		code = dict["code"] as String
		action = dict["action"] as? String
		if dict["editable"]
		{
			editable = dict["editable"] as Bool
		}
		
		if dict["title"]
		{
			title = dict["title"] as String
		}
		
		if dict["showDetail"]
		{
			showDetail = dict["showDetail"] as Bool
		}

		if dict["visible"]
		{
			visible = dict["visible"] as Bool
		}
		
		description = dict["description"] as? String
		image = dict["image"] as? String
	}
	
	func toDictionary() -> NSDictionary
	{
		var propertiesDictionary : NSMutableDictionary = NSMutableDictionary()
		propertiesDictionary.setValue(code, forKey: "code")
		propertiesDictionary.setValue(action, forKey: "action")
		propertiesDictionary.setValue(editable, forKey: "editable")
		propertiesDictionary.setValue(title, forKey: "title")
		propertiesDictionary.setValue(showDetail, forKey: "showDetail")
		propertiesDictionary.setValue(description, forKey: "description")
		propertiesDictionary.setValue(image, forKey: "image")
		propertiesDictionary.setValue(visible, forKey: "visible")
		return propertiesDictionary
	}
}